staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -0,0 +1,165 @@
+	declType, found := openAPISchemaTypes[s.Type]
+	if !found {
+		return nil
+	}
+
+	// We ignore XPreserveUnknownFields since we don't support validation rules on
+	// data that we don't have schema information for.
+
+	if s.XEmbeddedResource {
+		// 'apiVersion', 'kind', 'metadata.name' and 'metadata.generateName' are always accessible
+		// to validation rules since this part of the schema is well known and validated when CRDs
+		// are created and updated.
+		s = WithTypeAndObjectMeta(s)
+	}
+
+	switch declType.TypeName() {
+	case ListType.TypeName():
+		return NewListType(SchemaDeclType(s.Items))
+	case MapType.TypeName():
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			return NewMapType(StringType, SchemaDeclType(s.AdditionalProperties.Structural))
+		}
+		fields := make(map[string]*DeclField, len(s.Properties))
+
+		required := map[string]bool{}
+		if s.ValueValidation != nil {
+			for _, f := range s.ValueValidation.Required {
+				required[f] = true
+			}
+		}
+		for name, prop := range s.Properties {
+			var enumValues []interface{}
+			if prop.ValueValidation != nil {
+				for _, e := range prop.ValueValidation.Enum {
+					enumValues = append(enumValues, e.Object)
+				}
+			}
+			if fieldType := SchemaDeclType(&prop); fieldType != nil {
+				fields[Escape(name)] = &DeclField{
+					Name:         Escape(name),
+					Required:     required[name],
+					Type:         fieldType,
+					defaultValue: prop.Default.Object,
+					enumValues:   enumValues, // Enum values are represented as strings in CEL
+				}
+			}
+		}
+		return NewObjectType("object", fields)
+	case StringType.TypeName():
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "byte":
+				return StringType // OpenAPIv3 byte format represents base64 encoded string
+			case "binary":
+				return BytesType
+			case "duration":
+				return DurationType
+			case "date", "date-time":
+				return TimestampType
+			}
+		}
+	}
+	return declType
+}
+
+var (
+	openAPISchemaTypes = map[string]*DeclType{
+		"boolean":         BoolType,
+		"number":          DoubleType,
+		"integer":         IntType,
+		"null":            NullType,
+		"string":          StringType,
+		"date":            DateType,
+		"array":           ListType,
+		"object":          MapType,
+	}
+
+	// intOrStringType represents the x-kubernetes-int-or-string union type in CEL expressions.
+	// In CEL, the type is represented as an object where either the srtVal
+	// or intVal field is set. In CEL, this allows for typesafe expressions like:
+	//
+	// require that the string representation be a percentage:
+	//  `has(intOrStringField.strVal) && intOrStringField.strVal.matches(r'(\d+(\.\d+)?%)')`
+	// validate requirements on both the int and string representation:
+	//  `has(intOrStringField.intVal) ? intOrStringField.intVal < 5 : double(intOrStringField.strVal.replace('%', '')) < 0.5
+	//
+	intOrStringType = NewObjectType("intOrString", map[string]*DeclField{
```

Comment:
needs a TODO to reconcile with updates in https://github.com/kubernetes/enhancements/pull/3039